sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/NumberUtils.scala

```diff
@@ -0,0 +1,190 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.util
+
+import java.math.BigDecimal
+import java.text.{DecimalFormat, NumberFormat, ParsePosition}
+import java.util.Locale
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.errors.QueryExecutionErrors
+import org.apache.spark.sql.types.Decimal
+import org.apache.spark.unsafe.types.UTF8String
+
+object NumberUtils {
+
+  private val pointSign = '.'
+  private val letterPointSign = 'D'
+  private val commaSign = ','
+  private val letterCommaSign = 'G'
+  private val minusSign = '-'
+  private val letterMinusSign = 'S'
+  private val dollarSign = '$'
+
+  private def normalize(format: String): String = {
+    var notFindDecimalPoint = true
+    val normalizedFormat = format.toUpperCase(Locale.ROOT).map {
+      case '9' if notFindDecimalPoint => '#'
+      case '9' if !notFindDecimalPoint => '0'
+      case c if c == letterPointSign =>
+        notFindDecimalPoint = false
+        pointSign
+      case c if c == letterCommaSign => commaSign
+      case c if c == letterMinusSign => minusSign
+      case c if c == pointSign =>
+        notFindDecimalPoint = false
+        pointSign
+      case other => other
+    }
+    // If the comma is at the beginning or end of number format, then DecimalFormat will be invalid.
+    // For example, "##,###," or ",###,###" for DecimalFormat is invalid, so we must use "##,###"
+    // or "###,###".
+    UTF8String.fromString(normalizedFormat).trim(UTF8String.fromString(commaSign.toString)).toString
+  }
+
+  private def isSign(c: Char): Boolean = c match {
+    case c if c == pointSign || c == commaSign || c == minusSign || c == dollarSign => true
+    case _ => false
+  }
```

Comment:
nit (the set can be created in advance): 
```suggestion
  private def isSign(c: Char): Boolean = {
    Set(pointSign, commaSign, minusSign, dollarSign).contains(c)
  }
```